tensor_comprehensions/pybinds/tclib.cc

```diff
@@ -0,0 +1,498 @@
+#include <string>
+#include <vector>
+
+#include <glog/logging.h>
+
+#include <Python.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <torch/torch.h>
+
+#include "tc/aten/aten.h"
+#include "tc/aten/aten_autotuner.h"
+#include "tc/autotuner/genetic_search.h"
+#include "tc/autotuner/options_cache.h"
+#include "tc/core/cuda/cuda_backend.h"
+#include "tc/core/cuda/cuda_tc_executor.h"
+#include "tc/core/flags.h"
+#include "tc/core/tensor.h"
+#include "tc/lang/canonicalize.h"
+
+namespace tc {
+namespace python {
+
+namespace py = pybind11;
+
+namespace {
+void initGlog() {
+  static bool inited = false;
+  if (!inited) {
+    ::google::InitGoogleLogging("TC Python");
+    inited = true;
+  }
+}
+
+inline std::vector<at::Tensor> getATenTensors(py::tuple& pyTensors) {
+  std::vector<at::Tensor> atTensors;
+  for (auto& inp : pyTensors) {
+    atTensors.push_back(inp.cast<at::Tensor>());
+  }
+  return atTensors;
+}
+
+template <typename VoidPtr>
+inline std::vector<VoidPtr> getRawPtrsFromATenTensors(py::tuple& pyTensors) {
+  std::vector<VoidPtr> res;
+  for (auto& inp : pyTensors) {
+    res.push_back(static_cast<VoidPtr>(inp.cast<at::Tensor>().data_ptr()));
+  }
+  return res;
+}
+
+inline py::list convertToPyObjects(std::vector<at::Tensor>& tensors) {
+  py::list outputs;
+  for (auto& tensor : tensors) {
+    outputs.append(py::cast(torch::autograd::make_variable(tensor)));
+  }
+  return outputs;
+}
+
+} // namespace
+
+struct CudaCompilationUnit {
+  CudaCompilationUnit(const std::string& def) : tc(def) {
+    initGlog();
+  }
+  std::string tc;
+  std::unordered_map<std::string, std::unique_ptr<tc::CudaTcExecutor>> compiled;
+};
+
+using ATenCudaGeneticTuner =
+    tc::aten::ATenAutotuner<tc::CudaBackend, tc::autotune::GeneticSearch>;
+
+class CudaTuner : public ATenCudaGeneticTuner {
+ public:
+  CudaTuner(const std::string& tc, const std::string& cacheFileName = "")
+      : ATenCudaGeneticTuner(tc), fileName(cacheFileName) {}
+
+  std::string fileName;
+};
+
+class TunerConfig {
+ public:
+  TunerConfig(
+      uint32_t generations,
+      uint32_t populationSize,
+      uint32_t threads,
+      std::string devices,
+      bool logtostderr,
+      uint32_t stderrthreshold) {
+    generations_ = generations;
+    populationSize_ = populationSize;
+    threads_ = threads;
+    devices_ = devices;
+    logtostderr_ = logtostderr;
+    stderrthreshold_ = stderrthreshold;
+  }
+  // __enter__ / __exit__ in case we want to use a ContextManager in Python in
+  // the future. In any case, RAII and Python GC can just never work together.
+  void __enter__() const {
+    savedGenerations_ = tc::FLAGS_tuner_gen_generations;
+    savedPopulationSize_ = tc::FLAGS_tuner_gen_pop_size;
+    savedThreads_ = tc::FLAGS_tuner_threads;
+    savedDevices_ = tc::FLAGS_tuner_devices;
+    savedLogtostderr_ = FLAGS_logtostderr;
+    savedStderrthreshold_ = FLAGS_stderrthreshold;
+
+    tc::FLAGS_tuner_gen_generations = generations_;
```

Comment:
Maybe we should stop relying on flags aka global variables for configuration.  Not in this PR, obviously.